src/extensions/arithmetic.rs

```diff
@@ -0,0 +1,747 @@
+//! Basic arithmetic operations.
+
+use itertools::Itertools;
+use smol_str::SmolStr;
+
+use crate::{
+    resource::{ResourceSet, SignatureError},
+    types::{
+        type_param::{TypeArg, TypeArgError, TypeParam},
+        CustomType, HashableType, SimpleRow, SimpleType, TypeRow, TypeTag,
+    },
+    Resource,
+};
+
+use super::logic::bool_type;
+
+/// The resource identifier.
+pub const fn resource_id() -> SmolStr {
+    SmolStr::new_inline("Arithmetic")
+}
+
+const INT_PARAM: TypeParam = TypeParam::Value(HashableType::Int(8));
+
+fn int_type_id() -> SmolStr {
+    SmolStr::new_inline("int")
+}
+
+fn float64_type_id() -> SmolStr {
+    SmolStr::new_inline("float64")
+}
+
+fn int_type(n: u8) -> SimpleType {
+    CustomType::new(
+        int_type_id(),
+        [TypeArg::Int(n as u128)],
+        resource_id(),
+        TypeTag::Classic,
+    )
+    .into()
+}
+
+fn float64_type() -> SimpleType {
+    CustomType::new(float64_type_id(), [], resource_id(), TypeTag::Classic).into()
+}
+
+fn get_width(arg: &TypeArg) -> Result<u8, SignatureError> {
+    let n: u8 = match arg {
+        TypeArg::Int(n) => *n as u8,
+        _ => {
+            return Err(TypeArgError::TypeMismatch(arg.clone(), INT_PARAM).into());
+        }
+    };
+    if (n != 1)
+        && (n != 2)
+        && (n != 4)
+        && (n != 8)
+        && (n != 16)
+        && (n != 32)
+        && (n != 64)
+        && (n != 128)
```

Comment:
there is likely some bit mask thing for this but I think I prefer it readable like this anyway